File: 09_Trees/BFS and DFS/BFS_DFS.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <utility>
#include <cmath>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    
    TreeNode(int val) : val(val), left(nullptr), right(nullptr) {}
};

void printTree(TreeNode* root) {
    if (root == nullptr)
        return;
    
    queue<pair<TreeNode*, int>> q;
    q.push({root, 0});
    int prevLevel = 0;
    
    while (!q.empty()) {
        auto [node, level] = q.front();
        q.pop();
        
        if (level != prevLevel) {
            cout << endl;
            prevLevel = level;
        }
        
        int spaces = pow(2, 5 - level) - 1;
        cout << string(spaces, ' ') << node->val << string(spaces, ' ');
        
        if (node->left != nullptr)
            q.push({node->left, level + 1});
        if (node->right != nullptr)
            q.push({node->right, level + 1});
    }
    cout << endl;
}
void bfs(TreeNode* root) {
    if (!root) return;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        TreeNode* node = q.front();
        q.pop();
        cout << node->val << " ";
        if (node->left) q.push(node->left);
        if (node->right) q.push(node->right);
    }
    cout << endl;
}

void dfs(TreeNode* root) {
    if (!root) return;
    stack<TreeNode*> st;
    st.push(root);
    while (!st.empty()) {
        TreeNode* node = st.top();
        st.pop();
        cout << node->val << " ";
        if (node->right) st.push(node->right);
        if (node->left) st.push(node->left);
    }
    cout << endl;
}

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    cout << "Tree Structure:" << endl;
    printTree(root);

    cout << "BFS Traversal: ";
    bfs(root);

    cout << "DFS Traversal: ";
    dfs(root);

    return 0;
}

File: 09_Trees/BFS and DFS/BFS_DFS.java
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.Stack;


public class BFS_DFS {

    public static void printTree(TreeNode root) {
        if (root == null)
            return;
        
        Queue<Pair<TreeNode, Integer>> queue = new ArrayDeque<>();
        queue.add(new Pair<>(root, 0));
        int prevLevel = 0;
        
        while (!queue.isEmpty()) {
            Pair<TreeNode, Integer> pair = queue.poll();
            TreeNode node = pair.getKey();
            int level = pair.getValue();
            
            if (level != prevLevel) {
                System.out.println();
                prevLevel = level;
            }
            
            int spaces = (int) Math.pow(2, 5 - level) - 1;
            System.out.print(" ".repeat(spaces) + node.val + " ".repeat(spaces));
            
            if (node.left != null)
                queue.add(new Pair<>(node.left, level + 1));
            if (node.right != null)
                queue.add(new Pair<>(node.right, level + 1));
        }
        System.out.println();
        
    }

    public static void bfs(TreeNode root) {
        if (root == null)
            return;
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            System.out.print(node.val + " ");
            if (node.left != null)
                queue.add(node.left);
            if (node.right != null)
                queue.add(node.right);
        }
        System.out.println();
    }

    public static void dfs(TreeNode root) {
        if (root == null)
            return;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            System.out.print(node.val + " ");
            if (node.right != null)
                stack.push(node.right);
            if (node.left != null)
                stack.push(node.left);
        }
        System.out.println();
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);

        System.out.println("Tree Structure:");
        printTree(root);

        System.out.print("BFS Traversal: ");
        bfs(root);

        System.out.print("DFS Traversal: ");
        dfs(root);
    }
}

class TreeNode {
        int val;
        TreeNode left, right;
    
        public TreeNode(int val) {
            this.val = val;
            left = right = null;
        }
    }

    class Pair<F, S> {
        private F first;
        private S second;
    
        public Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }
    
        public F getKey() {
            return first;
        }
    
        public S getValue() {
            return second;
        }
    
        public void setFirst(F first) {
            this.first = first;
        }
    
        public void setSecond(S second) {
            this.second = second;
        }
    }
    